### PR TITLE
Fix/11972 documentation build with errors

### LINKED ISSUE #11972: Documentation builds with errors
### Issue Summary

Building the docs, while resulting in a working build, seems to raise some errors due to missing Python reference targets.

```
wagtail/docs/advanced_topics/api/v2/configuration.md:45: WARNING: py:class reference target not found: wagtail.api.v2.views.PagesAPIViewSet
wagtail/docs/advanced_topics/api/v2/configuration.md:46: WARNING: py:class reference target not found: wagtail.images.api.v2.views.ImagesAPIViewSet
wagtail/docs/advanced_topics/api/v2/configuration.md:47: WARNING: py:class reference target not found: wagtail.documents.api.v2.views.DocumentsAPIViewSet
wagtail/docs/advanced_topics/api/v2/configuration.md:48: WARNING: py:class reference target not found: wagtail.contrib.redirects.api.RedirectsAPIViewSet
wagtail/wagtail/documents/__init__.py:docstring of wagtail.documents.get_document_model:1: WARNING: py:class reference target not found: wagtail.documents.models.Document
wagtail/wagtail/images/__init__.py:docstring of wagtail.images.get_image_model:1: WARNING: py:class reference target not found: wagtail.images.models.Image
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.get_rendition:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.get_rendition:1: WARNING: py:class reference target not found: wagtail.images.models.AbstractRendition
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.find_existing_rendition:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.find_existing_rendition:1: WARNING: py:class reference target not found: wagtail.images.models.AbstractRendition
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.create_rendition:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.create_rendition:1: WARNING: py:class reference target not found: wagtail.images.models.AbstractRendition
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.get_renditions:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.get_renditions:1: WARNING: py:class reference target not found: wagtail.images.models.AbstractRendition
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.find_existing_renditions:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.find_existing_renditions:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.find_existing_renditions:1: WARNING: py:class reference target not found: wagtail.images.models.AbstractRendition
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.create_renditions:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.create_renditions:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.create_renditions:1: WARNING: py:class reference target not found: wagtail.images.models.AbstractRendition
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.generate_rendition_file:1: WARNING: py:class reference target not found: wagtail.images.models.Filter
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.generate_rendition_file:1: WARNING: py:class reference target not found: django.core.files.base.File
wagtail/wagtail/images/models.py:docstring of wagtail.images.models.AbstractImage.generate_rendition_file:1: WARNING: py:class reference target not found: django.core.files.base.File
wagtail/docs/extending/custom_account_settings.md:10: WARNING: py:class reference target not found: wagtail.users.models.UserProfile
wagtail/docs/extending/custom_account_settings.md:57: WARNING: py:class reference target not found: wagtail.users.models.UserProfile
wagtail/docs/reference/contrib/routablepage.md:: WARNING: py:attr reference target not found: wagtail.models.Page.url
wagtail/wagtail/models/__init__.py:docstring of wagtail.models.Page.route_for_request:1: WARNING: py:class reference target not found: HttpRequest
wagtail/wagtail/models/__init__.py:docstring of wagtail.models.Page.route_for_request:1: WARNING: py:class reference target not found: RouteResult
wagtail/wagtail/models/__init__.py:docstring of wagtail.models.Page.find_for_request:1: WARNING: py:class reference target not found: HttpRequest
wagtail/wagtail/models/__init__.py:docstring of wagtail.models.Page.update_aliases:1: WARNING: py:class reference target not found: PageRevision
wagtail/wagtail/models/__init__.py:docstring of wagtail.models.Page.update_aliases:1: WARNING: py:class reference target not found: optional
wagtail/docs/reference/pages/model_reference.md:497: WARNING: py:class reference target not found: wagtail.models.Locale
wagtail/docs/reference/pages/model_reference.md:784: WARNING: py:meth reference target not found: wagtail.models.RevisionsManager.for_instance
wagtail/docs/reference/pages/model_reference.md:1256: WARNING: py:meth reference target not found: LogEntryManger.log_action
wagtail/docs/reference/pages/panels.md:283: WARNING: py:class reference target not found: wagtail.contrib.forms.models.AbstractForm
wagtail/wagtail/blocks/field_block.py:docstring of wagtail.blocks.field_block.FieldBlock:1: WARNING: py:class reference target not found: wagtail.blocks.base.Block
wagtail/wagtail/blocks/field_block.py:docstring of wagtail.blocks.field_block.ChoiceBlock:1: WARNING: py:class reference target not found: wagtail.blocks.field_block.BaseChoiceBlock
wagtail/wagtail/blocks/field_block.py:docstring of wagtail.blocks.field_block.MultipleChoiceBlock:1: WARNING: py:class reference target not found: wagtail.blocks.field_block.BaseChoiceBlock
wagtail/wagtail/blocks/field_block.py:docstring of wagtail.blocks.field_block.PageChooserBlock:1: WARNING: py:class reference target not found: wagtail.blocks.field_block.ChooserBlock
wagtail/wagtail/documents/blocks.py:docstring of wagtail.documents.blocks.DocumentChooserBlock:1: WARNING: py:class reference target not found: wagtail.documents.views.chooser.BaseDocumentChooserBlock
wagtail/wagtail/images/blocks.py:docstring of wagtail.images.blocks.ImageChooserBlock:1: WARNING: py:class reference target not found: wagtail.blocks.field_block.ChooserBlock
wagtail/wagtail/snippets/blocks.py:docstring of wagtail.snippets.blocks.SnippetChooserBlock:1: WARNING: py:class reference target not found: wagtail.blocks.field_block.ChooserBlock
wagtail/wagtail/blocks/static_block.py:docstring of wagtail.blocks.static_block.StaticBlock:1: WARNING: py:class reference target not found: wagtail.blocks.base.Block
wagtail/wagtail/blocks/struct_block.py:docstring of wagtail.blocks.struct_block.StructBlock:1: WARNING: py:class reference target not found: wagtail.blocks.struct_block.BaseStructBlock
wagtail/wagtail/blocks/list_block.py:docstring of wagtail.blocks.list_block.ListBlock:1: WARNING: py:class reference target not found: wagtail.blocks.base.Block
wagtail/wagtail/blocks/stream_block.py:docstring of wagtail.blocks.stream_block.StreamBlock:1: WARNING: py:class reference target not found: wagtail.blocks.stream_block.BaseStreamBlock
wagtail/docs/reference/viewsets.md:67: WARNING: py:attr reference target not found: django.db.models.Options.model_name
wagtail/docs/releases/0.6.rst:31: WARNING: py:class reference target not found: wagtail.contrib.wagtailroutablepage.models.RoutablePageMixin
wagtail/docs/releases/1.0.rst:203: WARNING: py:class reference target not found: wagtail.contrib.wagtailroutablepage.models.RoutablePageMixin
wagtail/docs/releases/1.0.rst:203: WARNING: py:attr reference target not found: wagtail.contrib.wagtailroutablepage.models.RoutablePageMixin.subpage_urls
wagtail/docs/releases/1.1.rst:23: WARNING: py:mod reference target not found: wagtail.wagtailsearch
wagtail/docs/releases/1.1.rst:23: WARNING: py:mod reference target not found: wagtail.contrib.wagtailsearchpromotions
wagtail/docs/releases/1.1.rst:31: WARNING: py:mod reference target not found: wagtail.contrib.wagtailapi
wagtail/docs/releases/1.1.rst:103: WARNING: py:mod reference target not found: wagtail.wagtailsearch
wagtail/docs/releases/1.1.rst:108: WARNING: py:mod reference target not found: wagtail.contrib.wagtailsearchpromotions
wagtail/docs/releases/1.1.rst:119: WARNING: py:mod reference target not found: wagtail.contrib.wagtailsearchpromotions.models.SearchPromotion
wagtail/docs/releases/1.11.rst:124: WARNING: py:meth reference target not found: wagtail.core.models.Page.get_url_parts
wagtail/docs/releases/1.11.rst:124: WARNING: py:meth reference target not found: wagtail.core.models.Page.get_url
wagtail/docs/releases/1.11.rst:124: WARNING: py:meth reference target not found: wagtail.core.models.Page.get_full_url
wagtail/docs/releases/1.11.rst:124: WARNING: py:meth reference target not found: wagtail.core.models.Page.relative_url
wagtail/docs/releases/1.2.rst:118: WARNING: py:meth reference target not found: wagtail.core.models.Page.copy
wagtail/docs/releases/1.2.rst:187: WARNING: py:meth reference target not found: wagtail.core.models.Page.get_latest_revision_as_page
wagtail/docs/releases/1.6.1.rst:25: WARNING: py:meth reference target not found: wagtail.core.models.Page.can_create_at
wagtail/docs/releases/1.8.rst:73: WARNING: py:meth reference target not found: wagtail.core.query.PageQuerySet.in_site
wagtail/docs/releases/2.9.rst:97: WARNING: py:class reference target not found: wagtail.core.middleware.SiteMiddleware
wagtail/docs/releases/5.2.md:246: WARNING: py:attr reference target not found: wagtail.models.Revision.submitted_for_moderation
wagtail/docs/releases/5.2.md:247: WARNING: py:attr reference target not found: wagtail.models.Revision.submitted_revisions
wagtail/docs/releases/5.2.md:248: WARNING: py:meth reference target not found: wagtail.models.Revision.approve_moderation
wagtail/docs/releases/5.2.md:249: WARNING: py:meth reference target not found: wagtail.models.Revision.reject_moderation
wagtail/docs/topics/snippets/features.md:89: WARNING: py:class reference target not found: wagtail.search.index.Indexed
wagtail/docs/topics/snippets/features.md:329: WARNING: py:class reference target not found: taggit.manager.TaggableManager
wagtail/docs/topics/snippets/features.md:329: WARNING: py:class reference target not found: modelcluster.contrib.taggit.ClusterTaggableManager
```

### Steps to Reproduce

1. Follow the steps to build the docs https://docs.wagtail.org/en/stable/contributing/developing.html#compiling-the-documentation
2. Observe the console output

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?

-   The build should complete without errors.

### Working on this

Some of the missing references errors are due to a few reasons:
- Legacy code that has been removed, e.g. moderation methods in `wagtail.models.Revision`
- Wagtail's classes/functions that are undocumented (they don't have an entry in the reference docs)
- Classes/functions from external library that don't have an [intersphinx mapping](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) linked in our Sphinx config
- Potentially invalid markup in the docs

For legacy and undocumented references, we should just use normal inline code (<code>``</code>) instead of using the Sphinx reference directive (e.g. `{class}`). For external libraries, if they have an intersphinx object inventory, we should add them to our [`intersphinx_mapping`](https://github.com/wagtail/wagtail/blob/252bae9129f9167d371d65392ae95b2db1961de0/docs/conf.py#L134-L140) and fix the link as necessary. If they don't, either convert them to regular links or use normal inline code. For invalid markup, we should fix them.

Once we've resolved all the errors, we should look into making readthedocs fail loudly if the build completes with non-zero code exit code, so we can prevent new errors in the future. With our current config, we already get non-zero exit code as of #11503, but for some reason the RTD build is still marked as successful. We might need to [customise the build](https://docs.readthedocs.io/en/stable/build-customization.html) to get the desired behaviour.

Anyone can contribute to this. View our [contributing guidelines](https://docs.wagtail.org/en/latest/contributing/index.html), add a comment to the issue once you’re ready to start.

### PR DESCRIPTION
**Pull request by shentonfreude**
This builds upon HimanshuGarg47:fix/11972-documentation-build-with-errors, adding:

* silences warnings due to Sphinx not being able to resolve Wagtail type hint names (bug reports from 3 years back)
*  3 unresolved reference reportedly from "docstring" that are not  related to type hints (I can't tell what's causing them)
* one fix to allow Page.copy() to be resolved

fixes #11972 
replaces #12012 (commnitter, please credit @HimanshuGarg47 too when merging)

### CODE REVIEW

**Comment by laymonage** [comment]
Have briefly checked out the branch and confirmed that the build succeeds without any errors! 🎉 

image

Will do a full review later. Thank you very much!

**Comment by HimanshuGarg47** [comment]
Thanks, @shentonfreude, for fixing the other warnings. I was continuously struggling with the root cause unknown to the Sphinx in depth. I was supposed to fix the warning with the root cause, but as with your changes, it helped me to know that we can suppress the warnings and came to know about nitpick_ignore. I am happy that now there are no warnings.

**Comment by vossisboss** [comment]
I wasn't able to compile our documentation at the sprints at Wagtail Space US due to the following error after running `make html`:

```
Could not import extension sphinxcontrib.spelling (exception: The 'enchant' C library was not found and maybe needs to be installed.
See  https://pyenchant.github.io/pyenchant/install.html
for details
)
make: *** [html] Error 2
```

This might be a separate issue or I might be missing a dependency. Thought I'd mention it here in case it was relevant to this issue.

**Comment by shentonfreude** [comment]
I had the same problem in Arnhem and ended up creating a python docker
container with the required dependency

File: docs/reference/viewsets.md [hunk 1]
```diff
@@ -64,7 +64,7 @@ Viewsets are Wagtail's mechanism for defining a group of related admin views wit
 
    .. attribute:: model
 
-   Required; the model class that this viewset will work with. The :attr:`~django.db.models.Options.model_name` will be used
+   Required; the model class that this viewset will work with. The `~django.db.models.Options.model_name` will be used
```
Comment (gasman) [inline docs/reference/viewsets.md:67]: This renders as italics, which I don't think was intended - given that this reference doesn't resolve, I'll change this to a plain `model_name` (in double backticks).

File: docs/releases/0.6.rst [hunk 2]
```diff
@@ -28,7 +28,7 @@ Wagtail can now be used with Django 1.7.
 Minor features
 ~~~~~~~~~~~~~~
 * A new template tag has been added for reversing URLs inside routable pages. See :ref:`routablepageurl_template_tag`.
-* RoutablePage can now be used as a mixin. See :class:`wagtail.contrib.wagtailroutablepage.models.RoutablePageMixin`.
+* RoutablePage can now be used as a mixin. See `wagtail.contrib.wagtailroutablepage.models.RoutablePageMixin`.
```
Comment (gasman) [inline docs/releases/0.6.rst:31]: This renders as italics - changing to

    :ref:`routable_page_mixin`

**Comment by gasman** [comment]
Thanks @shentonfreude and @HimanshuGarg47! Merged in e066424163d7df7dc2c40b2e1c4d87fe03f6545d (with some remaining formatting errors fixed in a7d72f9fbf7d7e148da3e6a22dd469453dddfeea).

**Pull request closed without merging**